server/src/main/java/org/elasticsearch/rest/StreamingXContentResponse.java

```diff
@@ -0,0 +1,435 @@
+                    }
+                }
+            } finally {
+                queueRefs.decRef();
+            }
+        } else {
+            Releasables.closeExpectNoException(releasable);
+        }
+    }
+
+    /**
+     * A fragment which is ready for transmission, to be stored in {@link #fragmentQueue}.
+     *
+     * @param fragment      The fragment of XContent to send.
+     * @param releasable    A resource to release when this fragment has been fully transmitted, or is no longer required because the
+     *                      transmission was cancelled.
+     */
+    private record StreamingFragment(ChunkedToXContent fragment, Releasable releasable) {}
+
+    /**
+     * Queue of fragments that are ready for transmission.
+     */
+    private final Queue<StreamingFragment> fragmentQueue = new LinkedBlockingQueue<>();
+
+    /**
+     * Upper bound on the number of fragments in the queue, atomically modified to ensure there's only one thread processing the queue
+     * at once.
+     */
+    private final AtomicInteger queueLength = new AtomicInteger();
+
+    /**
+     * Ref-counting for access to the queue, to avoid clearing the queue on abort concurrently with a fragment being sent.
+     */
+    private final RefCounted queueRefs = AbstractRefCounted.of(this::drainQueue);
+
+    /**
+     * Flag to indicate if the request has been aborted, at which point we should stop enqueueing more fragments and promptly clean up the
+     * ones being sent. It's safe to ignore this, but without it in theory a constant stream of calls to {@link #writeFragment} could
+     * prevent {@link #drainQueue} from running for arbitrarily long.
+     */
+    private final AtomicBoolean isRestResponseFinished = new AtomicBoolean();
+
+    private boolean tryAcquireQueueRef() {
+        return isRestResponseFinished.get() == false && queueRefs.tryIncRef();
+    }
+
+    private void restResponseFinished() {
+        assert Transports.assertTransportThread();
+        if (isRestResponseFinished.compareAndSet(false, true)) {
+            queueRefs.decRef();
+        }
+    }
+
+    private void drainQueue() {
+        assert isRestResponseFinished.get();
+        assert queueRefs.hasReferences() == false;
+        final var taskCount = queueLength.get() + 2 /* currentFragmentReleasable and onCompletion */ ;
+        final var releasables = new ArrayList<Releasable>(taskCount);
+        try {
+            releasables.add(currentFragmentReleasable);
+            currentFragmentReleasable = null;
+            StreamingFragment fragment;
+            while ((fragment = fragmentQueue.poll()) != null) {
+                releasables.add(fragment.releasable());
+            }
+            assert fragmentQueue.isEmpty() : fragmentQueue.size(); // no concurrent adds
+            assert releasables.size() == taskCount - 1 || releasables.size() == taskCount - 2 : taskCount + " vs " + releasables.size();
```

Comment:
I'm missing something here, why would `releaseables.size() == taskCount - 2`, I get that `queueLength` is an upper bound, but this only happens after the enqueues have stopped, correct?

Reply:
We pop one item from the queue before we start to process it, but we decrement `queueLength` when its processing is complete. So mostly `queueLength` will be counting this one extra item that isn't really in the queue any more.